### PR TITLE
[Form] deprecate custom formats with HTML5 widgets

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#10986

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/DateType.php [hunk 1]
```diff
@@ -288,6 +288,14 @@ public function configureOptions(OptionsResolver $resolver)
         $resolver->setAllowedTypes('years', 'array');
         $resolver->setAllowedTypes('months', 'array');
         $resolver->setAllowedTypes('days', 'array');
+
+        $resolver->setNormalizer('html5', function (Options $options, $html5) {
+            if ($html5 && 'single_text' === $options['widget'] && self::HTML5_FORMAT !== $options['format']) {
+                @trigger_error(sprintf('Using a custom format when the "html5" option of the %s is enabled is deprecated since Symfony 4.2.', self::class), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/DateType.php]: will this code path throw in Symfony 5? or will we just ignore the property?
if throw => the code should be added as a commented line
if ignore => might end up being confusing

**Review by nicolas-grekas — COMMENTED** [review]
Why would it make Symfony better? I mean, yes, html5 is a standard, but why should we stricter than required actually? What's the benefit? I don't feel like this would make anything easier to maintain, nor easier to work with.

**Comment by xabbuh** [comment]
The thing is that custom formats do not work with `type="datetime-local` input fields. If you use a JavaScript-based widget you will hide this fact, but as soon as your application falls back to the HTML input for whatever reason, the input will not behave as expected. Using a custom format is totally fine, but in this case you need to opt-out of the HTML5 datetime input (i.e. setting `html5` to `false`). The rendering will then use a `type="text"` input which will behave as expected with custom formats.

File: src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php [hunk 2]
```diff
@@ -278,6 +282,15 @@ public function configureOptions(OptionsResolver $resolver)
             'text',
             'choice',
         ));
+
+        $resolver->setNormalizer('html5', function (Options $options, $html5) {
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php]: It needs rebase to use `setDeprecated()`, later the normalizer can throw the exception?

File: UPGRADE-4.2.md [hunk 3]
```diff
@@ -55,6 +55,7 @@ Finder
 Form
 ----
 
+ * Using the `format` option of the `DateType` and `DateTimeType` when the `html5` option is enabled is deprecated.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: 2nd "the" is strange when reading the sentence, isn't it? might miss a word, or should be removed? (same below)

**Comment by nicolas-grekas** [comment]
> custom formats do not work with type="datetime-local input fields

makes sense!

File: src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php [hunk 4]
```diff
@@ -278,6 +282,15 @@ public function configureOptions(OptionsResolver $resolver)
             'text',
             'choice',
         ));
+
+        $resolver->setDeprecated('html5', function (Options $options, $html5) {
+            if ($html5 && self::HTML5_FORMAT !== $options['format']) {
+                return sprintf('Using a custom format when the "html5" option of %s is enabled is deprecated since Symfony 4.3 and will lead to an exception in 5.0..', self::class);
```
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php]: double dot `..` after `5.0` (same below)

**Review by nicolas-grekas — APPROVED** [review]

**Pull request merged**